Framework/Algorithms/src/MaskNonOverlappingBins.cpp

```diff
@@ -0,0 +1,234 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidAlgorithms/MaskNonOverlappingBins.h"
+
+#include "MantidAPI/Axis.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidKernel/MultiThreaded.h"
+#include "MantidKernel/Unit.h"
+
+namespace {
+/// Constants for the algorithm's property names.
+namespace Prop {
+std::string const CHECK_SORTING{"CheckSortedX"};
+std::string const COMPARISON_WS{"ComparisonWorkspace"};
+std::string const INPUT_WS{"InputWorkspace"};
+std::string const MASK_PARTIAL{"MaskPartiallyOverlapping"};
+std::string const NONRAGGED{"SameXAcrossHistograms"};
+std::string const OUTPUT_WS{"OutputWorkspace"};
+} // namespace Prop
+
+/// Return true if X data is sorted in ascending order.
+bool isXSorted(Mantid::API::MatrixWorkspace const &ws) {
+  int unsorted{0};
+  PARALLEL_FOR_IF(Mantid::Kernel::threadSafe(ws))
+  for (int64_t i = 0; static_cast<size_t>(i) < ws.getNumberHistograms(); ++i) {
+    auto const &Xs = ws.x(static_cast<size_t>(i));
+    if (!std::is_sorted(Xs.cbegin(), Xs.cend())) {
+      PARALLEL_ATOMIC
+      ++unsorted;
+    }
+  }
```

Comment:
Why the int -> bool conversion? we count unsorted but throw the information away. Maybe you can clarify that. Otherwise looks great.

Reply:
Agree that `bool unsorted` would be more logical here, but one cannot assign `unsorted = true` in `PARALLEL_ATOMIC`; OpenMP does not support such constructs, at least on my system.